subsys/usb/usb_device.c

```diff
@@ -608,15 +608,29 @@ static bool usb_set_interface(u8_t iface, u8_t alt_setting)
 				if_desc = (void *)p;
 			}
 
-			LOG_DBG("iface_num %u alt_set %u", iface, alt_setting);
+			LOG_DBG("current iface %u current alt setting %u",
+				cur_iface, cur_alt_setting);
 			break;
 		case DESC_ENDPOINT:
-			if ((cur_iface != iface) ||
-			    (cur_alt_setting != alt_setting)) {
-				break;
+			if ((cur_iface == iface) && (cur_alt_setting != alt_setting)) {
+				/* disable endpoints from previous alternate */
+				LOG_DBG("Disable endpoint for iface:%d, alt_set:%d",
+					cur_iface, cur_alt_setting);
+				if (0 != usb_dc_ep_disable(((struct usb_ep_descriptor *)p)->bEndpointAddress)) {
+					LOG_ERR("Failed to disble endpoint for interface %d,"
+						"alt setting %d", cur_iface, cur_alt_setting);
+					found = false;
+				}
+			}
+			else if ((cur_iface == iface) && (cur_alt_setting == alt_setting)) {
+				LOG_DBG("Enable endpoint for interface %d, alt setting %d",
+					cur_iface, cur_alt_setting);
+				found = set_endpoint((struct usb_ep_descriptor *)p);
```

Comment:
Why not simply `usb_dc_ep_enable` when switching to alternate that has already been configured?

Reply:
The answer is that when we chose configuration (and associated endpoints) for the first time, there is a need to configure them. When switching to alternate interface there should not be that need. I will think about rework for this.
Endpoints should be configured once and then disabled/enabled depending on chosen alternate interface.

That was left from existing code when I reworked swtiching to alternate interfaces and didn't think about that in this matter.

This is a good point @masz-nordic, thanks! 🚀